stdlib/public/runtime/HeapObject.cpp

```diff
@@ -710,40 +710,42 @@ bool swift::isNativeSwiftWeakReference(WeakReference *ref) {
 }
 
 void swift::swift_weakInit(WeakReference *ref, HeapObject *value) {
-  ref->Value = (uintptr_t)value | WR_NATIVE;
+  ref->Value = (uintptr_t)value | (value ? WR_NATIVE : 0);
   SWIFT_RT_ENTRY_CALL(swift_unownedRetain)(value);
 }
 
 void swift::swift_weakAssign(WeakReference *ref, HeapObject *newValue) {
   SWIFT_RT_ENTRY_CALL(swift_unownedRetain)(newValue);
   auto oldValue = (HeapObject*) (ref->Value & ~WR_NATIVE);
-  ref->Value = (uintptr_t)newValue | WR_NATIVE;
+  ref->Value = (uintptr_t)newValue | (newValue ? WR_NATIVE : 0);
   SWIFT_RT_ENTRY_CALL(swift_unownedRelease)(oldValue);
 }
 
 HeapObject *swift::swift_weakLoadStrong(WeakReference *ref) {
-  if (ref->Value == (uintptr_t)nullptr) {
-    return nullptr;
-  }
-
   // ref might be visible to other threads
-  auto ptr = __atomic_fetch_or(&ref->Value, WR_READING, __ATOMIC_RELAXED);
-  while (ptr & WR_READING) {
-    short c = 0;
-    while (__atomic_load_n(&ref->Value, __ATOMIC_RELAXED) & WR_READING) {
-      if (++c == WR_SPINLIMIT) {
-        std::this_thread::yield();
-        c -= 1;
-      }
+  auto ptr = ref->Value;
+  short c = 0;
+  while (true) {
+    // Never write WR_READING into a null pointer.
+    // This would cause a race against swift_unknownWeakLoadStrong.
+    if (ptr == (uintptr_t)nullptr)
+      return nullptr;
+    // Set WR_READING if it is not set already.
+    if (!(ptr & WR_READING)) {
+      if (__atomic_compare_exchange_n(&ref->Value, &ptr, ptr|WR_READING, false,
+                                      __ATOMIC_RELAXED, __ATOMIC_RELAXED))
+        break;
+      // Failed to set WR_READING. atomic_compare_exchange loaded ptr again.
+    }
+    // Failed to set WR_READING or WR_READING is already set. Retry.
+    if (++c == WR_SPINLIMIT) {
+      std::this_thread::yield();
+      c -= 1;
+      ptr = ref->Value;
```

Comment:
Do this load and the one above the loop need to be done with __atomic_load_n?  It probably wouldn't hurt.